src/app/Http/Controllers/Operations/ShowOperation.php

```diff
@@ -70,18 +70,43 @@ public function show($id)
 
         // get entry ID from Request (makes sure its the last ID for nested resources)
         $id = $this->crud->getCurrentEntryId() ?? $id;
-        $setFromDb = $this->crud->get('show.setFromDb');
 
         // get the info for that entry
         $this->data['entry'] = $this->crud->getEntry($id);
         $this->data['crud'] = $this->crud;
         $this->data['title'] = $this->crud->getTitle() ?? trans('backpack::crud.preview').' '.$this->crud->entity_name;
 
-        // set columns from db
-        if ($setFromDb) {
+        // load the view from /resources/views/vendor/backpack/crud/ if it exists, otherwise load the one in the package
+        return view($this->crud->getShowView(), $this->data);
+    }
+
+    /**
+     * Default behaviour for setupShowOperation, in case none has been
+     * provided by including a setupShowOperation() method in the CrudController.
+     */
+    protected function setupShowOperation()
+    {
+        // guess which columns to show, from the database table
+        if ($this->crud->get('show.setFromDb')) {
             $this->crud->setFromDb(false, true);
         }
 
+        // if the developer has chosen to include a modifyShowOperation() method
+        // in his/her CrudController, make sure to call it because they probably want to
+        // add a new column or remove an autoset column
+        if (method_exists($this, 'modifyShowOperation')) {
+            $this->modifyShowOperation();
```

Comment:
We don't have this behavior in any other operations, and I think it also does not make sense since the `setupShowOperation` could now be defined by developer. 

My rationale is: if you write `public function modifyShowOperation()` why not `public function setupShowOperation()`? 

What I guess will happen is:

```php
public function setupShowOperation() {
     $this->setupListOperation();
}
```
What am missing here? 

Pedro

Reply:
Good catch. This inconsistency with other operations is something that I feel uncomfortable about too. And that's what I expect most _new_ people to do too - call `$this->setupListOperation()`.

But there's quite a big difference between that and what `modifyShowOperation()` does. Say you have 4 columns added to `list`, out of total 10 columns in the db table. Then:
- if you do nothing, you don't define `setupShowOperation()`, you'll have all 10 columns in your show;
- if you do `setupShowOperation() { $this->setupListOperation(); }` you'll have 4 columns in your Show; those other ones... are lost, you have to define them manually or call `$this->crud->setFromDb()` yourself;
- if you do `modifyShowOperation() {}`, inside there all 10 columns have been added, you can change them to different types or remove them; 

Again, I'm not very happy with this different behaviour either. But it's what I want most of the time from the Show operation - show me not only the columns in the List, but all the columns in the db. Then, what I want is to change column types. That's how I find myself using it.

One way to look at it is that:
- because List is limited in space (each column takes up width), by default no columns are present;
- because Show is unlimited in space (you can easily scroll down), by default all columns are present;

That's the way it's functioned so far, the default behaviour for Show has been to show all columns in the db. We _could_ change this default behaviour, but... I don't think we should. It's useful. It gives you 90% of what you want, most of the time.

Reply:
Let me give this another try, I think I got where you are going, but I think I didn't explain myself perfectly clear: 

I still think that method is not needed. If you add columns they don't delete the previous columns, atmost it overrides the ones with same names (or it should). 

I think if it's "default" show operation behaviour, it should be in `setupShowOperationDefaults`, leaving the `setupShowOperation` to be fully controled by developer. 

Wouldn't that make it similar in behavior with all other operations and still keep the ability to "modify" whatever you want inside `setupShowOperation` ? 

Let me know,
Pedro

Reply:
I agree - it would make A LOT more sense in `setupShowOperationDefaults()`, in terms of where code goes. That would be so clean! But the problem is... the order they're run in:
- first `setupShowOperationDefaults()`
- then `setupShowOperation()`

That means if we do all the adding and removing of columns in `setupShowOperationDefaults()`... there's no way for the dev to easily turn that off, from `setupShowOperation()`. It's kind of an irreversible behaviour. 

Even if we do it `setupShowOperationDefaults()` in a conditional (`if (CRUD::get('show.autoset')) {}`)... the dev cannot change that config _before_ everything happens... only afterwards...

If I remember correctly, that's the reason I ended up with `modifyShowOperation()`... that and (of course) upgradability.

Reply:
Isn't possible: 
```php
if(method_exists(setupShowOperation)) { dont run please } ?
```
We just need to tell devs that want to customize the operation to manually call their initial setup ? 

```php
public function setupShowOperation() {
     $this->setupShowOperationDefaults();
}
```
We are telling them to use `modifyShowOperation` if they want to achieve the same: - use default, modify after so ... why just don't tell them a different text, but that makes more sense, atleast for me:
- NO setupOperation: setupDefaults is run.
- setupOperation: Defaults are not run if not specified by dev because they are taking control of setup.

```php
public function setupShowOperation() {
    $this->setupShowOperationDefaults();
    $this->removeColumn('xxx');
}
```

If you are worried about "consistency" about operation, we can introduce an operation feature that is: `run defaults on load or allways run defaults`, so operations like create etc run the defaults even when `setupCreateOperation` is present. But it would benefit operations like this one, that would have `run defaults on load => false` that would make it not run defaults when developer takes control of operation setup. 

PS: forget the name, even I reading this see it's completely wrong, but I hope you got my idead.

Reply:
AAHAHAHAH!!! 😅

![stupid_me_3](https://user-images.githubusercontent.com/1032474/145450841-9407ab27-ac84-479d-90f0-f5616c736bb5.gif)

You're totally right @pxpm - let me try it out, we should be able to use `if(method_exists(setupShowOperation))` for this. We can't call `setupShowOperationDefaults()` because that does extra things too, and will be called no matter what, but we can just as well create a new function with the default logic, and call that. So it could be something like:

```php
// no setupShowOperation() defined
// setupShowOperationDefaults() will run guessColumnsForShowOperation()

// otherwise
public function setupShowOperation() {
    $this->guessColumnsForShowOperation();
    $this->removeColumn('xxx');
}
```

For the upgrader then, it's just as easy. We don't tell them "change `setupShowOperation()` to `modifyShowOperation()`, we tell them:

```diff
public function setupShowOperation() {
+    $this->guessColumnsForShowOperation();
      // your existing logic
```

Which is just as good (if not better). I LOVE IT. Thanks a lot @pxpm , great idea!

Reply:
Done! Changed it, great catch @pxpm .